### PR TITLE
Using Let's Encrypt certificates in Java applications

### PR DESCRIPTION
**Pull request by kencoenen**

### CODE REVIEW

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+When you want to enable HTTPS on your website or need certificates for TLS communication, you'll need to request this certificate from a [Certificate Authority](https://en.wikipedia.org/wiki/Certificate_authority){:target="_blank" rel="noopener noreferrer"} (CA).
+It acts as a [trusted third party](https://en.wikipedia.org/wiki/Trusted_third_party){:target="_blank" rel="noopener noreferrer"} between two parties that need to communicate with each other.
+[Let's Encrypt](https://certbot.eff.org/){:target="_blank" rel="noopener noreferrer"} is such a Certificate Authority.
+It is their mission to give everyone a secure and privacy-respecting web experience.
+That's why they issue certificates free of charge.
+
+# Installing a Let's Encrypt certificate
+
+We assume that you have shell access to your server.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Suggestion: Assuming that you have shell access to your server, Let's Encrypt recommends ...

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+
+Certbot is a free, open source software tool for automatically using Let’s Encrypt certificates on manually-administrated websites to enable HTTPS.
+
+On the Certbot website you can find clear instructions on how to install.
+You can select your web server software (Apache, Nginx, ...) and operating system and Certbot provides the instructions to install it.
+
+> You can check your operating system on Linux by executing `cat /etc/os-release`.
+
+Please note that these instructions also include setting up HTTPS for your website, which for this tutorial isn't necessary.
+We'll use the certificate in another way, for TLS communication in a Java applications.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: application — no S.

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+
+For Ubuntu, these are the steps required to install Certbot.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Suggestion: ... , the following steps are required ...

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+See also [Apache on Ubuntu 16.04 (xenial)](https://certbot.eff.org/lets-encrypt/ubuntuxenial-apache){:target="_blank" rel="noopener noreferrer"}.
+
+You'll need to add the Certbot PPA to your list of repositories.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Suggestion: PPA stands for? Add acronym tag?

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+sudo apt-get update
+```
+
+Run this command on the command line on the machine to install Certbot.
+
+```
+sudo apt-get install certbot python-certbot-apache
+```
+
+You can choose to retrieve a certificate, or retrieve it and install it immediately in your web server, eg. Apache.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: *on* your web server, ...

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+```
+
+In the script above, we also write the log information to a file for debugging purposes.
+You can also add the `-q` parameter for quiet execution, but then writing to a log file has no use anymore.
+
+You can add this shell script as a Cronjob on your server. 
+The command to renew certbot is installed in one of the following locations: `/etc/crontab/`, `/etc/cron.*/*` or `systemctl list-timers`.
+eg. To execute the script once every hour, you can add it to `/etc/cron.hourly`.
+
+We will edit the contents of the crontab file on the system with ``crontab -e``.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Double back-ticks?

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+.---------------- minute (0 - 59)
+| .------------- hour (0 - 23)
+| | .---------- day of month (1 - 31)
+| | | .------- month (1 - 12) OR jan,feb,mar,apr ...
+| | | | .---- day of week (0 - 6) (Sunday=0 or 7) OR sun,mon,tue,wed,thu,fri,sat
+| | | | |
+* * * * * command to be executed
+```
+
+You can read more on cron expressions in the Baeldung blog [A Guide To Cron Expressions](https://www.baeldung.com/cron-expressions).
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Open in new tab

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+```
+
+Remember to set execute permissions on the created script to allow the system to run the script.
+
+```
+chmod +x /home/<username>/renew-script.sh
+```
+
+There's no need to restart something after changing the `crontab` file.
+Cron will examine the modification time on all crontabs and reload those which have changed.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: which have changes *or* which contain changes ?

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+
+Let's Encrypt certificates are stored in the `/etc/letsencrypt/live` folder on your file system.
+We will now see how we can import them in Java keystore files to use them in a Java application.
+
+## Importing certificates into `cacerts`
+
+The first way you can use certificates in a JVM is to add them to the `cacerts` file of your Java distribution.
+
+Every JRE has its own keystore, which contains all Certificate Authorities it trusts.
+This is also refered to as a `truststore`.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: referred

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+This `truststore` is stored as file called `cacerts`.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: as *a* file

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+
+```
+keytool -import -alias mydomain.be -keystore $JAVA_HOME/jre/lib/security/cacerts -file /etc/letsencrypt/live/mydomain.be/cert.pem -storepass changeit -noprompt
+```
+
+Please note that adding certificates to `cacerts` is not always the best solution.
+Although technically it is a fully functional keystore file, its purpose is mainly for determining which third-party certificates to trust.
+On top of this, it is tied to your Java installation and when you install another JRE or JDK, you'll need to add the certificates again.
+
+Our preferred approach is to add your own certificates to a keystore and the third-party certificates to separate truststore.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: to *a* separate

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+The next step is to import the certificates into a `.keystore` file.
+
+```
+keytool -importkeystore -deststorepass changeit -destkeypass changeit -deststoretype pkcs12 -srckeystore /tmp/mydomain.be.p12 -srcstoretype PKCS12 -srcstorepass changeit -destkeystore /tmp/mydomain.be.keystore -alias mydomain.be
+```
+
+You can now load the keystore at location `/tmp/mydomain.be.keystore` in your Java application.
+
+> Please note that you not only need to create a keystore with your own certificates, but also a truststore with the trusted third-party certificates.
+The approach is exactly the same however.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Suggestion: However, the approach is exactly the same.

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+We create the `CloseableHttpClient` with the `org.apache.http.impl.client.HttpClients` utility class.
+
+```
+final CloseableHttpClient client = HttpClients
+    .custom()
+    .setSSLContext(sslContext)
+    .build();
+```
+
+> If you don't have the `.setSSLContext(sslContext)`, please check your ``org.apache.httpcomponents:httpclient`` version.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Double back-ticks?

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+
+You can now execute requests on with that client and each request will be sent over a TLS connection.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: on that client *or* with that client ?

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+We'll use `HttpComponentsClientHttpRequestFactory` to customize the `RestTemplate` and we register it with the Spring context by annotating the method with `@Bean` in one of the configuration classes.
+
+```
+@Bean
+public RestTemplate restTemplate() throws Exception {
+    return new RestTemplate(new HttpComponentsClientHttpRequestFactory(client));
+}
+```
+
+Please note that in the above code snippet, ``client`` must be an instance of Apache's HttpComponents HttpClient, eg. the `CloseableHttpClient` we created in the previous section.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Double back-ticks?

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+    String url = properties.getRemoteServiceEndpointUrl();
+    client.setDefaultUri(url);
+    client.setMarshaller(marshaller);
+    client.setUnmarshaller(marshaller);
+    client.setMessageSender(new HttpComponentsMessageSender(client));
+    return client;
+}
+```
+
+Like in the Spring REST example, ``client`` must be an instance of Apache's HttpComponents HttpClient, eg. the `CloseableHttpClient` we created earlier.
```
Comment (MrJean) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Double back-ticks?

**Comment by bjorndecraemer** [comment]
Looks good, except for the spelling errors that Jan has already identified.
What is unclear for me as a reader is that is describes how to automatically renew your certificate, but not how to adapt the keystore when a new cert is added.

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+authors: [ken_coenen]
+title: 'Using Let''s Encrypt certificates in Java applications'
+image: /img/lets-encrypt.png
+tags: [Security, TLS, Let's Encrypt, Keycloak]
+category: Security
+comments: true
+---
+
+> At some point in their career, developers come accross the need to work with security certificates.
+This article describes how to setup Let's Encrypt, retrieve a certificate, renew it automatically and how to use the certificate in a Java application for TLS communication.
```
Comment (JMesens) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: omit thelast  'how to' before 'use the certificate'

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+It is their mission to give everyone a secure and privacy-respecting web experience.
+That's why they issue certificates free of charge.
+
+# Installing a Let's Encrypt certificate
+
+Assuming that you have shell access to your server, Let's Encrypt recommends to use [Certbot ACME Client](https://certbot.eff.org/){:target="_blank" rel="noopener noreferrer"}, since it can automate certificate issuance and installation with zero downtime.
+
+Certbot is a free, open source software tool for automatically using Let’s Encrypt certificates on manually-administrated websites to enable HTTPS.
+
+On the Certbot website you can find clear instructions on how to install.
```
Comment (JMesens) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: on how to instal 'it'; or rewrite something as: You can find clear installation instructions on the Certbot website.

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+You can select your web server software (Apache, Nginx, ...) and operating system and Certbot provides the instructions to install it.
+
+> You can check your operating system on Linux by executing `cat /etc/os-release`.
+
+Please note that these instructions also include setting up HTTPS for your website, which for this tutorial isn't necessary.
+We'll use the certificate in another way, for TLS communication in a Java application.
+
+For Ubuntu, the following steps are required to install Certbot.
+See also [Apache on Ubuntu 16.04 (xenial)](https://certbot.eff.org/lets-encrypt/ubuntuxenial-apache){:target="_blank" rel="noopener noreferrer"}.
```
Comment (JMesens) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Also see

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+
+Certbot is installed using APT (Advanced Package Tool), a tool for installing and removing applications on your system. This tool searches in its repositories for software distributions. Before you can install Certbot, you'll need to add the Certbot PPA (Personal Package Archive) to your list of available APT repositories.
```
Comment (JMesens) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: 'installing and removing applications on Debian based systems' (minor)

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+sudo apt-get update
+```
+
+Run the following command to install Certbot.
+
+```
+sudo apt-get install certbot python-certbot-apache
+```
+
+You can choose to retrieve a certificate, or retrieve it and install it immediately on your web server, eg. Apache by adding the parameter `--apache`.
```
Comment (JMesens) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Maybe split up in multiple sentences for better readability.

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 22]
```diff
@@ -0,0 +1,409 @@
...
+
+Your application must provide a bean instance of the type `WebServiceGatewaySupport`.
+On this object, you can set an instance of a `WebServiceMessageSender`.
+
+
+```
+@Bean
+public Jaxb2Marshaller marshaller() {
+    Jaxb2Marshaller marshaller = new Jaxb2Marshaller();
+    marshaller.setContextPath("be.belgacom.pgf.services.provisioningapisyncrequest.v3");
```
Comment (JMesens) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Aren't they named pxs now? ^^

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+It's even easier, as you don't need to write code to read the `.keystore` file.
+You can point to the `.keystore` file in the configuration files for that product.
+
+This is an example Keycloak configuration in the `standalone.xml` file:
+
+```
+<security-realm name="ApplicationRealm">
+    <server-identities>
+        <ssl>
+            <keystore path="/tmp/api.connectmy.car.keystore" relative-to="jboss.server.config.dir" keystore-password="changeit" alias="keycloak.ordina-jworks.io" key-password="changeit" generate-self-signed-certificate-host="localhost"/>
```
Comment (JMesens) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: connectmy copy paste, replace maybe?

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 24]
```diff
@@ -0,0 +1,409 @@
...
+For our situation, it is enough to retrieve a certificate.
+You do this with the following command:
+
+```
+certbot certonly
+```
+
+You can find the installed Let's Encrypt certificates in the `/etc/letsencrypt/live` folder on your file system.
+
+# Certificate renewal
```
Comment (jmeys) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Certbot [automatically renews](https://certbot.eff.org/docs/using.html?highlight=hooks#automated-renewals) certificates on most operating systems now.
you can verify this by looking at `systemctl list-timers`.

```
NEXT                         LEFT        LAST                         PASSED              UNIT                         ACTIVATES
Wed 2019-08-14 10:37:47 UTC  15h left    n/a                          n/a                 certbot.timer                certbot.service
```

Testing the renewal is a good practice though. Example output of `certbot renew --dry-run`:
```
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processing /etc/letsencrypt/renewal/mydomain.be.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Cert not due for renewal, but simulating renewal for dry run
Plugins selected: Authenticator standalone, Installer None
Renewing an existing certificate

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
new certificate deployed without reload, fullchain is
/etc/letsencrypt/live/mydomain.be/fullchain.pem
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
** DRY RUN: simulating 'certbot renew' close to cert expiry
**          (The test certificates below have not been saved.)

Congratulations, all renewals succeeded. The following certs have been renewed:
  /etc/letsencrypt/live/mydomain.be/fullchain.pem (success)
** DRY RUN: simulating 'certbot renew' close to cert expiry
**          (The test certificates above have not been saved.)
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 25]
```diff
@@ -0,0 +1,409 @@
...
+To use a certificate in a Java application, the preferred way is to add it to a `.keystore` file.
+
+The Java Runtime Environment (JRE) ships with a tool called `keytool` to create certificates and manipulate key stores.
+Adding certificates to a keystore can be done by using OpenSSL and the `keytool`.
+
+You cannot import `.pem` certificates directly in a keystore, so you'll first need to add the `.pem` file to a [PKCS 12](https://en.wikipedia.org/wiki/PKCS_12){:target="_blank" rel="noopener noreferrer"} archive.
+We do this with the OpenSSL tool with the following command.
+
+```
+openssl pkcs12 -export -in /etc/letsencrypt/live/mydomain.be/cert.pem -inkey /etc/letsencrypt/live/mydomain.be/privkey.pem -out /tmp/mydomain.be.p12 -name mydomain.be -CAfile /etc/letsencrypt/live/mydomain.be/fullchain.pem -caname "Let's Encrypt Authority X3" -password pass:changeit
```
Comment (jmeys) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Suggestion to split this line for readability:
```
openssl pkcs12 -export \
	 -in /etc/letsencrypt/live/mydomain.be/cert.pem \
	 -inkey /etc/letsencrypt/live/mydomain.be/privkey.pem \
	 -out /tmp/mydomain.be.p12 \
	 -name mydomain.be \
	 -CAfile /etc/letsencrypt/live/mydomain.be/fullchain.pem \
	 -caname "Let's Encrypt Authority X3" \
	 -password pass:changeit
```

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 26]
```diff
@@ -0,0 +1,409 @@
...
+```
+
+Change `mydomain.be` with your own DNS name.
+
+The next step is to import the certificates into a `.keystore` file.
+
+```
+keytool -importkeystore -deststorepass changeit -destkeypass changeit -deststoretype pkcs12 -srckeystore /tmp/mydomain.be.p12 -srcstoretype PKCS12 -srcstorepass changeit -destkeystore /tmp/mydomain.be.keystore -alias mydomain.be
```
Comment (jmeys) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Suggestion to split this line for readability:
```
keytool -importkeystore \
	-deststorepass changeit \
	-destkeypass changeit \
	-deststoretype pkcs12 \
	-srckeystore /tmp/mydomain.be.p12 \
	-srcstoretype PKCS12 \
	-srcstorepass changeit \
	-destkeystore /tmp/mydomain.be.keystore \
	-alias mydomain.be
```

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 27]
```diff
@@ -0,0 +1,409 @@
...
+Every JRE has its own keystore, which contains all Certificate Authorities it trusts.
+This is also referred to as a `truststore`.
+This `truststore` is stored as a file called `cacerts`.
+It is typically located in `$JAVA_HOME/jre/lib/security` assuming `$JAVA_HOME` is where your JRE or JDK is installed.
+The default password for this keystore is `changeit`.
+
+The following command imports the certificates into your JRE truststore.
+
+```
+keytool -import -alias mydomain.be -keystore $JAVA_HOME/jre/lib/security/cacerts -file /etc/letsencrypt/live/mydomain.be/cert.pem -storepass changeit -noprompt
```
Comment (jmeys) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Suggestion to split this line for readability:
```
keytool -import -alias mydomain.be \
	-keystore $JAVA_HOME/jre/lib/security/cacerts \
	-file /etc/letsencrypt/live/mydomain.be/cert.pem \
	-storepass changeit \
	-noprompt
```

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md (hunk 19 above)
Comment (kencoenen) [reply _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: "See also..." is correct, as is "You can also see" and "Also, see..." but not "Also see...".

https://forum.wordreference.com/threads/see-also-vs-also-see.2092638/

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md (hunk 24 above)
Comment (kencoenen) [reply _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Nice catch, thanks! 2y ago that wasn't the case.
I'll rewrite this part and also mention the automatic cron setup.

**Comment by kencoenen** [comment]
> Looks good, except for the spelling errors that Jan has already identified.
> What is unclear for me as a reader is that is describes how to automatically renew your certificate, but not how to adapt the keystore when a new cert is added.

Moved the cronjob part to after the section to create a keystore to automate the keystore creation.

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 28]
```diff
@@ -0,0 +1,505 @@
...
+
+Assuming that you have shell access to your server, Let's Encrypt recommends to use [Certbot ACME Client](https://certbot.eff.org/){:target="_blank" rel="noopener noreferrer"}, since it can automate certificate issuance and installation with zero downtime.
+
+Certbot is a free, open source software tool for automatically using Let’s Encrypt certificates on manually-administrated websites to enable HTTPS.
+
+You can find clear installation instructions on the Certbot website.
+You can select your web server software (Apache, Nginx, ...) and operating system and Certbot provides the instructions to install it.
+
+> You can check your operating system on Linux by executing `cat /etc/os-release`.
+
```
Comment (KevinVHoutte) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: (You can) three times a charm, try rephrasing it in to something more fluent for the readers

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 29]
```diff
@@ -0,0 +1,505 @@
...
+	-destkeystore /tmp/mydomain.be.keystore \
+	-alias mydomain.be
+```
+
+You can now load the keystore at location `/tmp/mydomain.be.keystore` in your Java application.
+
+> Please note that you not only need to create a keystore with your own certificates, but also a truststore with the trusted third-party certificates.
+However, the approach is exactly the same.
+
+## Automate the keytore and truststore creation process
```
Comment (KevinVHoutte) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]:
```suggestion
## Automate the keystore and truststore creation process
```

File: _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md [hunk 30]
```diff
@@ -0,0 +1,505 @@
...
+```
+final CloseableHttpClient client = HttpClients
+    .custom()
+    .setSSLContext(sslContext)
+    .build();
+```
+
+> If you don't have the `.setSSLContext(sslContext)`, please check your `org.apache.httpcomponents:httpclient` version.
+
+Each HTTP request executed using this client will be sent over a TLS connection.
```
Comment (KevinVHoutte) [inline _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]:
```suggestion
Each HTTP request executed using this client, will be sent over a TLS connection.
```

Comment (kencoenen) [reply _posts/2019-08-13-Using-Lets-Encrypt-Certificates-In-Java.md]: Better without comma

**Pull request merged**